src/Database/LSMTree/Internal/Unsliced.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveFunctor     #-}
+{-# LANGUAGE FlexibleInstances #-}
+
+module Database.LSMTree.Internal.Unsliced (
+    -- * Unsliced raw bytes
+    Unsliced
+    -- * Unsliced keys
+  , makeUnslicedKey
+  , unsafeMakeUnslicedKey
+  , fromUnslicedKey
+  , unsafeNoAssertMakeUnslicedKey
+  ) where
+
+import           Control.Exception (assert)
+import qualified Data.Vector.Primitive as PV
+import           Data.Word (Word8)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey (..))
+import           Database.LSMTree.Internal.Serialise.RawBytes (RawBytes (..))
+import           Database.LSMTree.Internal.Vector (noRetainedExtraMemory)
+
+-- | A type @a@ represented as unsliced raw bytes.
+newtype Unsliced a =
+    UnsafeNoAssertUnsliced (PV.Vector Word8)
```

Comment:
So this is essentially `ByteArray`? We could save two machine words if we don't use `Vector` here.